From 463b03920c12e4d61b1fb1ba0d4515a95290ffbb Mon Sep 17 00:00:00 2001
From: Rick Chen <rick@andestech.com>
Date: Thu, 29 Jun 2023 14:59:22 +0800
Subject: [PATCH 086/191] riscv: andes: Fix WARN_ON when PROVE_LOCK is enabled

When Kernel execute patch function feature, it was expected
to lock the text_mutex. By default PROVE_LOCK is not enabled, it
will ignore the lock checking mechanism. As PROVE_LOCK is enabled,
it will check if text_mutex is locked, otherwise there comes the
WARN_ON messages during booting progress. Add to lock text_mutex
can fix this issue.

Signed-off-by: Rick Chen <rick@andestech.com>
---
 arch/riscv/errata/andes/errata.c | 7 ++++++-
 arch/riscv/kernel/cpufeature.c   | 2 ++
 2 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/arch/riscv/errata/andes/errata.c b/arch/riscv/errata/andes/errata.c
index 0442ccaf46e1..da28f685085e 100644
--- a/arch/riscv/errata/andes/errata.c
+++ b/arch/riscv/errata/andes/errata.c
@@ -14,6 +14,8 @@
 #include <soc/andes/sbi.h>
 #include <soc/andes/andes.h>
 #include <soc/andes/ppma.h>
+#include <linux/memory.h>
+#include <linux/mutex.h>
 
 bool andes_legacy_mmu;
 EXPORT_SYMBOL(andes_legacy_mmu);
@@ -123,7 +125,10 @@ void __init_or_module andes_errata_patch_func(struct alt_entry *begin,
 			continue;
 
 		tmp = (1U << alt->errata_id);
-		if (cpu_req_errata & tmp)
+		if (cpu_req_errata & tmp) {
+			mutex_lock(&text_mutex);
 			patch_text_nosync(alt->old_ptr, alt->alt_ptr, alt->alt_len);
+			mutex_unlock(&text_mutex);
+		}
 	}
 }
diff --git a/arch/riscv/kernel/cpufeature.c b/arch/riscv/kernel/cpufeature.c
index 58bf81a5ef66..64b7b70b4d8c 100644
--- a/arch/riscv/kernel/cpufeature.c
+++ b/arch/riscv/kernel/cpufeature.c
@@ -21,6 +21,8 @@
 #include <asm/processor.h>
 #include <asm/smp.h>
 #include <asm/switch_to.h>
+#include <linux/memory.h>
+#include <linux/mutex.h>
 
 #define NUM_ALPHA_EXTS ('z' - 'a' + 1)
 
-- 
2.34.1

